File: app/src/main/java/com/commit451/gitlab/activity/ProjectActivity.java
package com.commit451.gitlab.activity;

import android.content.ClipData;
import android.content.ClipboardManager;
import android.content.Context;
import android.content.Intent;
import android.os.Bundle;
import android.support.annotation.NonNull;
import android.support.design.widget.Snackbar;
import android.support.design.widget.TabLayout;
import android.support.v4.app.Fragment;
import android.support.v4.view.ViewPager;
import android.support.v7.widget.Toolbar;
import android.view.MenuItem;
import android.view.View;
import android.view.ViewGroup;
import android.widget.Toast;

import com.commit451.alakazam.HideRunnable;
import com.commit451.gitlab.App;
import com.commit451.gitlab.R;
import com.commit451.gitlab.adapter.ProjectSectionsPagerAdapter;
import com.commit451.gitlab.event.ProjectReloadEvent;
import com.commit451.gitlab.fragment.BaseFragment;
import com.commit451.gitlab.model.Ref;
import com.commit451.gitlab.model.api.Project;
import com.commit451.gitlab.navigation.Navigator;
import com.commit451.gitlab.rx.CustomSingleObserver;
import com.commit451.gitlab.util.IntentUtil;

import org.parceler.Parcels;

import butterknife.BindView;
import butterknife.ButterKnife;
import io.reactivex.Single;
import io.reactivex.android.schedulers.AndroidSchedulers;
import io.reactivex.schedulers.Schedulers;
import timber.log.Timber;

public class ProjectActivity extends BaseActivity {

    private static final String EXTRA_PROJECT = "extra_project";
    private static final String EXTRA_PROJECT_ID = "extra_project_id";
    private static final String EXTRA_PROJECT_NAMESPACE = "extra_project_namespace";
    private static final String EXTRA_PROJECT_NAME = "extra_project_name";

    private static final String STATE_REF = "ref";
    private static final String STATE_PROJECT = "project";

    private static final int REQUEST_BRANCH_OR_TAG = 1;

    public static Intent newIntent(Context context, Project project) {
        Intent intent = new Intent(context, ProjectActivity.class);
        intent.putExtra(EXTRA_PROJECT, Parcels.wrap(project));
        return intent;
    }

    public static Intent newIntent(Context context, String projectId) {
        Intent intent = new Intent(context, ProjectActivity.class);
        intent.putExtra(EXTRA_PROJECT_ID, projectId);
        return intent;
    }

    public static Intent newIntent(Context context, String projectNamespace, String projectName) {
        Intent intent = new Intent(context, ProjectActivity.class);
        intent.putExtra(EXTRA_PROJECT_NAMESPACE, projectNamespace);
        intent.putExtra(EXTRA_PROJECT_NAME, projectName);
        return intent;
    }

    @BindView(R.id.root)
    ViewGroup root;
    @BindView(R.id.toolbar)
    Toolbar toolbar;
    @BindView(R.id.tabs)
    TabLayout tabLayout;
    @BindView(R.id.progress)
    View progress;
    @BindView(R.id.pager)
    ViewPager viewPager;

    Project project;
    Ref ref;

    private final Toolbar.OnMenuItemClickListener onMenuItemClickListener = new Toolbar.OnMenuItemClickListener() {
        @Override
        public boolean onMenuItemClick(MenuItem item) {
            switch (item.getItemId()) {
                case R.id.action_branch:
                    if (project != null) {
                        Navigator.navigateToPickBranchOrTag(ProjectActivity.this, project.getId(), ref, REQUEST_BRANCH_OR_TAG);
                    }
                    return true;
                case R.id.action_share:
                    if (project != null) {
                        IntentUtil.share(root, project.getWebUrl());
                    }
                    return true;
                case R.id.action_copy_git_https:
                    if (project == null || project.getHttpUrlToRepo() == null) {
                        Toast.makeText(ProjectActivity.this, R.string.failed_to_copy_to_clipboard, Toast.LENGTH_SHORT)
                                .show();
                    } else {
                        copyToClipboard(project.getHttpUrlToRepo());
                    }
                    return true;
                case R.id.action_copy_git_ssh:
                    if (project == null || project.getHttpUrlToRepo() == null) {
                        Toast.makeText(ProjectActivity.this, R.string.failed_to_copy_to_clipboard, Toast.LENGTH_SHORT)
                                .show();
                    } else {
                        copyToClipboard(project.getSshUrlToRepo());
                    }
                    return true;
            }
            return false;
        }
    };

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_project);
        ButterKnife.bind(this);
        Project project = Parcels.unwrap(getIntent().getParcelableExtra(EXTRA_PROJECT));

        if (savedInstanceState != null) {
            project = Parcels.unwrap(savedInstanceState.getParcelable(STATE_PROJECT));
            ref = Parcels.unwrap(savedInstanceState.getParcelable(STATE_REF));
        }
        toolbar.setNavigationIcon(R.drawable.ic_back_24dp);
        toolbar.setNavigationOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                finish();
            }
        });
        toolbar.inflateMenu(R.menu.menu_project);
        toolbar.setOnMenuItemClickListener(onMenuItemClickListener);

        if (project == null) {
            String projectId = getIntent().getStringExtra(EXTRA_PROJECT_ID);
            String projectNamespace = getIntent().getStringExtra(EXTRA_PROJECT_NAMESPACE);
            if (projectId != null) {
                loadProject(projectId);
            } else if (projectNamespace != null) {
                String projectName = getIntent().getStringExtra(EXTRA_PROJECT_NAME);
                loadProject(projectNamespace, projectName);
            } else {
                throw new IllegalStateException("You did something wrong and now we don't know what project to load. :(");
            }
        } else {
            bindProject(project);
        }
    }

    @Override
    protected void onActivityResult(int requestCode, int resultCode, Intent data) {
        super.onActivityResult(requestCode, resultCode, data);
        switch (requestCode) {
            case REQUEST_BRANCH_OR_TAG:
                if (resultCode == RESULT_OK) {
                    ref = Parcels.unwrap(data.getParcelableExtra(PickBranchOrTagActivity.EXTRA_REF));
                    broadcastLoad();
                }
                break;
        }
    }

    @Override
    protected void onSaveInstanceState(Bundle outState) {
        super.onSaveInstanceState(outState);
        outState.putParcelable(STATE_REF, Parcels.wrap(ref));
        outState.putParcelable(STATE_PROJECT, Parcels.wrap(project));
    }

    private void loadProject(String projectId) {
        showProgress();
        loadProject(App.get().getGitLab().getProject(projectId));
    }

    private void loadProject(String projectNamespace, String projectName) {
        showProgress();
        loadProject(App.get().getGitLab().getProject(projectNamespace, projectName));
    }

    private void loadProject(Single<Project> observable) {
        observable.compose(this.<Project>bindToLifecycle())
                .subscribeOn(Schedulers.io())
                .observeOn(AndroidSchedulers.mainThread())
                .subscribe(new CustomSingleObserver<Project>() {

                    @Override
                    public void error(@NonNull Throwable t) {
                        Timber.e(t);
                        progress.animate()
                                .alpha(0.0f)
                                .withEndAction(new HideRunnable(progress));
                        Snackbar.make(root, getString(R.string.connection_error), Snackbar.LENGTH_SHORT)
                                .show();
                    }

                    @Override
                    public void success(@NonNull Project project) {
                        progress.animate()
                                .alpha(0.0f)
                                .withEndAction(new HideRunnable(progress));
                        bindProject(project);
                    }
                });
    }

    private void showProgress() {
        progress.setAlpha(0.0f);
        progress.setVisibility(View.VISIBLE);
        progress.animate().alpha(1.0f);
    }

    private void broadcastLoad() {
        App.bus().post(new ProjectReloadEvent(project, ref.getRef()));
    }

    @Override
    public void onBackPressed() {
        Fragment fragment = getSupportFragmentManager().findFragmentByTag("android:switcher:" + R.id.pager + ":" + viewPager.getCurrentItem());
        if (fragment instanceof BaseFragment) {
            if (((BaseFragment) fragment).onBackPressed()) {
                return;
            }
        }

        super.onBackPressed();
    }

    public String getRef() {
        if (ref == null) {
            return null;
        }
        return ref.getRef();
    }

    public Project getProject() {
        return project;
    }

    private void bindProject(Project project) {
        this.project = project;
        if (ref == null) {
            ref = new Ref(Ref.TYPE_BRANCH, this.project.getDefaultBranch());
        }
        toolbar.setTitle(this.project.getName());
        toolbar.setSubtitle(this.project.getNamespace().getName());
        setupTabs();
    }

    private void setupTabs() {
        ProjectSectionsPagerAdapter projectSectionsPagerAdapter = new ProjectSectionsPagerAdapter(this, getSupportFragmentManager());
        viewPager.setAdapter(projectSectionsPagerAdapter);
        tabLayout.setupWithViewPager(viewPager);
    }

    private void copyToClipboard(String url) {
        ClipboardManager clipboard = (ClipboardManager) getSystemService(Context.CLIPBOARD_SERVICE);

        // Creates a new text clip to put on the clipboard
        ClipData clip = ClipData.newPlainText(project.getName(), url);
        clipboard.setPrimaryClip(clip);
        Snackbar.make(root, R.string.copied_to_clipboard, Snackbar.LENGTH_SHORT)
                .show();
    }
}

File: app/src/main/java/com/commit451/gitlab/activity/ProjectActivity.kt
package com.commit451.gitlab.activity

import android.app.Activity
import android.content.ClipData
import android.content.ClipboardManager
import android.content.Context
import android.content.Intent
import android.net.Uri
import android.os.Bundle
import android.view.View
import android.view.ViewGroup
import android.widget.Toast
import androidx.appcompat.widget.Toolbar
import androidx.viewpager.widget.ViewPager
import butterknife.BindView
import butterknife.ButterKnife
import com.commit451.addendum.extraOrNull
import com.commit451.alakazam.fadeOut
import com.commit451.gitlab.App
import com.commit451.gitlab.R
import com.commit451.gitlab.adapter.ProjectPagerAdapter
import com.commit451.gitlab.data.Prefs
import com.commit451.gitlab.event.ProjectReloadEvent
import com.commit451.gitlab.extension.with
import com.commit451.gitlab.fragment.BaseFragment
import com.commit451.gitlab.model.Ref
import com.commit451.gitlab.model.api.Project
import com.commit451.gitlab.navigation.DeepLinker
import com.commit451.gitlab.navigation.Navigator
import com.commit451.gitlab.util.IntentUtil
import com.google.android.material.snackbar.Snackbar
import com.google.android.material.tabs.TabLayout
import io.reactivex.Single
import timber.log.Timber

class ProjectActivity : BaseActivity() {

    companion object {

        private const val EXTRA_PROJECT = "extra_project"
        private const val EXTRA_PROJECT_ID = "extra_project_id"
        private const val EXTRA_PROJECT_NAMESPACE = "extra_project_namespace"
        private const val EXTRA_PROJECT_NAME = "extra_project_name"
        private const val EXTRA_PROJECT_SELECTION = "extra_project_selection"

        private const val STATE_REF = "ref"
        private const val STATE_PROJECT = "project"

        private const val REQUEST_BRANCH_OR_TAG = 1

        fun newIntent(context: Context, project: Project): Intent {
            val intent = Intent(context, ProjectActivity::class.java)
            intent.putExtra(EXTRA_PROJECT, project)
            return intent
        }

        fun newIntent(context: Context, projectId: String): Intent {
            val intent = Intent(context, ProjectActivity::class.java)
            intent.putExtra(EXTRA_PROJECT_ID, projectId)
            return intent
        }

        fun newIntent(context: Context, projectNamespace: String, projectName: String, projectSelection: DeepLinker.ProjectSelection): Intent {
            val intent = Intent(context, ProjectActivity::class.java)
            intent.putExtra(EXTRA_PROJECT_NAMESPACE, projectNamespace)
            intent.putExtra(EXTRA_PROJECT_NAME, projectName)
            intent.putExtra(EXTRA_PROJECT_SELECTION, projectSelection)
            return intent
        }
    }

    @BindView(R.id.root)
    lateinit var root: ViewGroup
    @BindView(R.id.toolbar)
    lateinit var toolbar: Toolbar
    @BindView(R.id.tabs)
    lateinit var tabLayout: TabLayout
    @BindView(R.id.progress)
    lateinit var progress: View
    @BindView(R.id.pager)
    lateinit var viewPager: ViewPager

    var project: Project? = null
    var ref: Ref? = null

    private var adapter: ProjectPagerAdapter? = null

    private val projectSelection by extraOrNull<DeepLinker.ProjectSelection>(EXTRA_PROJECT_SELECTION)

    private val onMenuItemClickListener = Toolbar.OnMenuItemClickListener { item ->
        when (item.itemId) {
            R.id.action_branch -> {
                if (project != null) {
                    Navigator.navigateToPickBranchOrTag(this@ProjectActivity, project!!.id, ref, REQUEST_BRANCH_OR_TAG)
                }
                return@OnMenuItemClickListener true
            }
            R.id.action_share -> {
                if (project != null) {
                    IntentUtil.share(root, Uri.parse(project!!.webUrl))
                }
                return@OnMenuItemClickListener true
            }
            R.id.action_copy_git_https -> {
                val url = project?.httpUrlToRepo
                if (url == null) {
                    Toast.makeText(this@ProjectActivity, R.string.failed_to_copy_to_clipboard, Toast.LENGTH_SHORT)
                            .show()
                } else {
                    copyToClipboard(url)
                }
                return@OnMenuItemClickListener true
            }
            R.id.action_copy_git_ssh -> {
                val url = project?.sshUrlToRepo
                if (url == null) {
                    Toast.makeText(this@ProjectActivity, R.string.failed_to_copy_to_clipboard, Toast.LENGTH_SHORT)
                            .show()
                } else {
                    copyToClipboard(url)
                }
                return@OnMenuItemClickListener true
            }
        }
        false
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        Prefs.startingView = Prefs.STARTING_VIEW_PROJECTS
        setContentView(R.layout.activity_project)
        ButterKnife.bind(this)
        var project: Project? = intent.getParcelableExtra(EXTRA_PROJECT)

        if (savedInstanceState != null) {
            project = savedInstanceState.getParcelable(STATE_PROJECT)
            ref = savedInstanceState.getParcelable(STATE_REF)
        }
        toolbar.setNavigationIcon(R.drawable.ic_back_24dp)
        toolbar.setNavigationOnClickListener { finish() }
        toolbar.inflateMenu(R.menu.menu_project)
        toolbar.setOnMenuItemClickListener(onMenuItemClickListener)

        if (project == null) {
            val projectId = intent.getStringExtra(EXTRA_PROJECT_ID)
            val projectNamespace = intent.getStringExtra(EXTRA_PROJECT_NAMESPACE)
            if (projectId != null) {
                loadProject(projectId)
            } else if (projectNamespace != null) {
                val projectName = intent.getStringExtra(EXTRA_PROJECT_NAME)
                loadProject(projectNamespace, projectName)
            } else {
                throw IllegalStateException("You did something wrong and now we don't know what project to load. :(")
            }
        } else {
            bindProject(project)
        }
    }

    override fun onActivityResult(requestCode: Int, resultCode: Int, data: Intent?) {
        super.onActivityResult(requestCode, resultCode, data)
        when (requestCode) {
            REQUEST_BRANCH_OR_TAG -> if (resultCode == Activity.RESULT_OK) {
                ref = data?.getParcelableExtra(PickBranchOrTagActivity.EXTRA_REF)
                broadcastLoad()
            }
        }
    }

    override fun onSaveInstanceState(outState: Bundle) {
        super.onSaveInstanceState(outState)
        outState.putParcelable(STATE_REF, ref)
        outState.putParcelable(STATE_PROJECT, project)
    }

    override fun onBackPressed() {
        val fragment = supportFragmentManager.findFragmentByTag("android:switcher:" + R.id.pager + ":" + viewPager.currentItem)
        if (fragment is BaseFragment) {
            if (fragment.onBackPressed()) {
                return
            }
        }

        super.onBackPressed()
    }

    override fun hasBrowsableLinks(): Boolean {
        return true
    }

    override fun onNewIntent(intent: Intent?) {
        super.onNewIntent(intent)
        val selection = intent?.getSerializableExtra(EXTRA_PROJECT_SELECTION) as? DeepLinker.ProjectSelection
        selection?.let {
            val index = adapter?.indexForSelection(it)
            if (index != null) {
                viewPager.setCurrentItem(index, false)
            }
        }
    }

    private fun loadProject(projectId: String) {
        showProgress()
        loadProject(App.get().gitLab.getProject(projectId))
    }

    private fun loadProject(projectNamespace: String, projectName: String) {
        showProgress()
        loadProject(App.get().gitLab.getProject(projectNamespace, projectName))
    }

    private fun loadProject(observable: Single<Project>) {
        observable.with(this)
                .subscribe({
                    progress.fadeOut()
                    bindProject(it)
                }, {
                    Timber.e(it)
                    progress.fadeOut()
                    Snackbar.make(root, getString(R.string.connection_error), Snackbar.LENGTH_INDEFINITE)
                            .show()
                })
    }

    private fun showProgress() {
        progress.alpha = 0.0f
        progress.visibility = View.VISIBLE
        progress.animate().alpha(1.0f)
    }

    private fun broadcastLoad() {
        App.bus().post(ProjectReloadEvent(project!!, ref!!.ref!!))
    }

    fun getRefRef(): String? {
        if (ref == null) {
            return null
        }
        return ref!!.ref
    }

    private fun bindProject(project: Project) {
        this.project = project
        if (ref == null) {
            ref = Ref(Ref.TYPE_BRANCH, project.defaultBranch)
        }
        toolbar.title = project.name
        toolbar.subtitle = project.namespace.name
        setupTabs()
    }

    private fun setupTabs() {
        val adapter = ProjectPagerAdapter(this, supportFragmentManager)
        this.adapter = adapter
        viewPager.adapter = adapter
        tabLayout.setupWithViewPager(viewPager)
        projectSelection?.let {
            val index = adapter.indexForSelection(it)
            viewPager.setCurrentItem(index, false)
        }
    }

    private fun copyToClipboard(url: String) {
        val clipboard = getSystemService(Context.CLIPBOARD_SERVICE) as ClipboardManager

        // Creates a new text clip to put on the clipboard
        val clip = ClipData.newPlainText(project!!.name, url)
        clipboard.setPrimaryClip(clip)
        Snackbar.make(root, R.string.copied_to_clipboard, Snackbar.LENGTH_SHORT)
                .show()
    }
}
